src/btree/bt_sync.c

```diff
@@ -299,9 +299,13 @@ __sync_file(WT_SESSION_IMPL *session, WT_CACHE_OP syncop)
              * cache clean but with history that cannot be
              * discarded), that is not wasted effort because
              * checkpoint doesn't need to write the page again.
+             *
+             * Once the transaction has given up it's snapshot it
+             * is no longer safe to reconcile pages. That happens
+             * prior to the final metadata checkpoint.
              */
             if (!WT_PAGE_IS_INTERNAL(page) && page->read_gen == WT_READGEN_WONT_NEED &&
-              !tried_eviction) {
+              F_ISSET(&session->txn, WT_TXN_HAS_SNAPSHOT) && !tried_eviction) {
```

Comment:
WT-4965 modified these conditions but backport to 4.0 for it was rejected. I think these changes here are correct, but it will be worthwhile to confirm with @agorrod once again if WT-4965 is not needed and these conditions here are complete for 4.0.

Reply:
This particular change looks correct to me. I still believe WT-5050 isn't relevant for 4.0, since it was only for situations where the stable and oldest timestamps diverge, which doesn't happen until MongoDB 4.2.